compiler/cpp/test/lexer/LexerTest.cpp

```diff
@@ -0,0 +1,435 @@
+
+TEST(LexerTests, TestLexerAdvancedHTML) {
+    FileReader reader = FileReader("test-files");
+    std::string advancedInput = reader.fileToString("advanced", "advanced.html");
+    Lexer lexer = Lexer(false, advancedInput, "", "<!--", "-->");
+
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_BLOCK_IDEN_OPEN);
+    lexer.expect(TOK_IF);
+    lexer.expect(TOK_HAS);
+    lexer.expect(TOK_NOT);
+    lexer.expect(TOK_TRUE);
+    lexer.expect(TOK_FALSE);
+    lexer.expect(TOK_OR);
+    lexer.expect(TOK_EQUALS);
+    lexer.expect(TOK_NOT_EQUALS);
+    lexer.expect(TOK_LESS);
+    lexer.expect(TOK_LESS_EQUAL);
+    lexer.expect(TOK_NUMBER);
+    lexer.expect(TOK_GREATER);
+    lexer.expect(TOK_GREATER_EQUAL);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_NUMBER);
+    lexer.expect(TOK_STRING);
+    lexer.expect(TOK_NUMBER);
+    lexer.expect(TOK_DOT);
+    lexer.expect(TOK_INDEX);
+    lexer.expect(TOK_BRACE_OPEN);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_BRACE_CLOSE);
+    lexer.expect(TOK_COM_BLOCK_IDEN_CLOSE);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_BLOCK_IDEN_OPEN);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_STRING);
+    lexer.expect(TOK_BRACE_OPEN);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_BRACE_CLOSE);
+    lexer.expect(TOK_COM_BLOCK_IDEN_CLOSE);
+    lexer.expect(TOK_ARBITRARY);
+}
+
+TEST(LexerTests, TestLexerAdvancedYAML) {
+    FileReader reader = FileReader("test-files");
+    std::string advancedInput = reader.fileToString("advanced", "advanced.yml");
+    Lexer lexer = Lexer(false, advancedInput, "#", "", "");
+
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_LINE_IDEN);
+    lexer.expect(TOK_IF);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_DOT);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_INDEX);
+    lexer.expect(TOK_DOT);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_EQUALS);
+    lexer.expect(TOK_FALSE);
+    lexer.expect(TOK_BRACE_OPEN);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_LINE_IDEN);
+    lexer.expect(TOK_BRACE_CLOSE);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_LINE_IDEN);
+    lexer.expect(TOK_IF);
+    lexer.expect(TOK_HAS);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_DOT);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_BRACE_OPEN);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_LINE_IDEN);
+    lexer.expect(TOK_BRACE_CLOSE);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_LINE_IDEN);
+    lexer.expect(TOK_IF);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_DOT);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_INDEX);
+    lexer.expect(TOK_DOT);
+    lexer.expect(TOK_IDENTIFIER);
+    lexer.expect(TOK_NOT_EQUALS);
+    lexer.expect(TOK_STRING);
+    lexer.expect(TOK_BRACE_OPEN);
+    lexer.expect(TOK_ARBITRARY);
+    lexer.expect(TOK_COM_LINE_IDEN);
+    lexer.expect(TOK_BRACE_CLOSE);
+}
+
```

Comment:
I know it's a real pain, but you should test not only the token type but also the values of these tokens. It's ugly I know and it also means to write down parts of the file in your code to test against with all new lines and other pitfalls. Take a look at the Java tests for the advanced files (Java and HTML). You should also test that the token has the correct line number. I know this takes a lot of effort but I think it's worth it in the end.

Reply:
Will tackle that, but it is going to take a while ...

Reply:
Took me three hours, but it is ready now. All tests now contain value checks and code pos checks 😅

Reply:
awesome work 👏